src/openms/source/ANALYSIS/QUANTITATION/IsotopeLabelingMDVs.cpp

```diff
@@ -0,0 +1,322 @@
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Ahmed Khalil $
+// $Authors: Ahmed Khalil $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/ANALYSIS/QUANTITATION/IsotopeLabelingMDVs.h>
+
+namespace OpenMS
+{
+
+  const std::string IsotopeLabelingMDVs::NamesOfFeatureName[] = {"intensity", "peak_apex_int"};
+
+  IsotopeLabelingMDVs::IsotopeLabelingMDVs() :
+    DefaultParamHandler("IsotopeLabelingMDVs")
+  {
+  }
+
+  IsotopeLabelingMDVs::~IsotopeLabelingMDVs()
+  {
+  }
+
+  void IsotopeLabelingMDVs::updateMembers_()
+  {
+  }
+  
+  void IsotopeLabelingMDVs::isotopicCorrection(
+    const Feature& normalized_feature,
+    Feature& corrected_feature,
+    const Matrix<double>& correction_matrix,
+    const DerivatizationAgent& correction_matrix_agent)
+  {
+    // MDV_corrected = correction_matrix_inversed * MDV_observed (normalized_features)
+    
+    /// Correction Matrices for various derivatization agents
+    const std::map<DerivatizationAgent, std::vector<std::vector<double>> > correction_matrices_ =
+    {
+      { DerivatizationAgent::TBDMS, {{0.8213, 0.1053, 0.0734, 0.0000},
+                                     {0.8420, 0.0963, 0.0617, 0.0000},
+                                     {0.8466, 0.0957, 0.0343, 0.0233},
+                                     {0.8484, 0.0954, 0.0337, 0.0225}}
+      }
+    };
+    
+    Eigen::MatrixXd correction_matrix_eigen;
+    auto correction_matrix_search = correction_matrices_.find(correction_matrix_agent);
+    
+    // use the internally saved correction matrix if the derivatization agent name is supplied and found
+    if (correction_matrix_agent != DerivatizationAgent::NOT_SELECTED && correction_matrix_search != correction_matrices_.end())
+    {
+      correction_matrix_eigen.resize(correction_matrix_search->second.size(), correction_matrix_search->second[0].size());
```

Comment:
OpenMS already contains code which can do the conversion:
https://abibuilder.informatik.uni-tuebingen.de/archive/openms/Documentation/nightly/html/namespaceOpenMS.html#a51eb708c742e3091feadac71be8bb50f
unfortunately, someone decided to allocate the EigenMatrix on the heap .. for whatever reason.
So the interface is inefficient.
A better one would be to just return the matrix on the stack.
If you want, feel free to add another function to MatrixUtils.h, otherwise I can also do that, since it's not really a part of this PR.